### PR TITLE
Refactor conference participant handling and phone number normalization

### PR DESCRIPTION
**Pull request by farmanahmed888**
# Pull Request

##  Summary
- Refactor participant handling to improve phone number normalization and error handling. Remove auto-mute functionality for students on join and enhance participant selection logic. Update API service for consistent phone number formatting across conference operations.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### High-Level Review Summary

The code has been refactored to normalize phone numbers at API boundaries, which improves consistency. Several enhancements include batch processing for adding participants, better handling of 'raised hand' status, and improved user selection logic based on normalized phone comparisons. There are improvements for robustness and user experience, but some issues need addressing:

### Issues & Improvement Areas
1. **Inconsistent Phone Normalization**: 
   - Phone numbers are normalized in API services and utility functions, but some parts (e.g., conference SSE event handling, user toggles, and conference students filtering) still use raw or inconsistent formats. Ensure all parts handle normalized data uniformly.

2. **Type Safety & Robustness**:
   - Defensive checks in `ParticipantCard`, `conference students` merging, and context toggle functions are good, but some variables can still be `undefined`, leading to potential runtime errors.
   - Some optional chaining (`?.`) is used but inconsistent; consider TypeScript or PropTypes for stricter type safety.

3. **Complex Merging Logic**:
   - The merge of conference student data with real-time user data is overly complicated. It can be simplified by adopting a unified state model, for example, tracking all participants in one normalized structure.

4. **Duplicated Logic & Redundant Code**:
   - Multiple phone normalization calls and phone number comparisons throughout the code; centralize normalization in utility functions and consistently use normalized data.
   - API calls (mute/unmute/add participant) normalize phone twice. Avoid double normalization.

5. **UI/UX Minor**:
   - Button labels are inconsistent; the "Sink conference" seems mislabeled, replaced with "End conference" which is correct.
   - Feedback in `ConferenceContext` logs raised hand status but lacks in the UI representation.

### Design / Performance
- Batch-add participants for API efficacy — good.
- Handling race conditions when toggling participants or updating conference state can be further improved with immutable patterns or state management libraries.

### Final Scoring (out of 100)
**75/100**

The code demonstrates good incremental improvements, but needs stricter type safety, centralized normalization, and simplified data merging logic to meet higher standards of robustness and maintainability. Consider integrating TypeScript and focusing on uniform data handling for excellence.

**Recommended for merging after addressing normalization consistency, code deduplication, and improved type safety.**

File: teacher-webapp/src/components/participants/ParticipantCard.js [hunk 1]
```diff
@@ -74,7 +79,7 @@ export const ParticipantCard = ({
           </Typography>
           <Box sx={{ display: "flex", alignItems: "center", gap: 0.5, mt: 0.5 }}>
             <Typography variant="body2" color="text.secondary">
-              {participant.phoneNumber}
+              {participant.phoneNumber || "N/A"}
```
Comment (Soumabha) [inline teacher-webapp/src/components/participants/ParticipantCard.js]: Participant should always have phone number as is deemed during student and teacher addition. Defensive checks will just hide symptoms - not a fix.

File: teacher-webapp/src/components/participants/ParticipantCard.js [hunk 2]
```diff
@@ -102,8 +107,8 @@ export const ParticipantCard = ({
         {canReconnect && (
           <Tooltip title="Reconnect">
             <IconButton
-              onClick={() => onReconnect(participant.phoneNumber)}
-              disabled={isReconnecting}
+              onClick={() => onReconnect && onReconnect(participant.phoneNumber)}
+              disabled={isReconnecting || !participant.phoneNumber}
```
Comment (Soumabha) [inline teacher-webapp/src/components/participants/ParticipantCard.js]: Not clear on why the participant.phoneNumber check is necessary. We should maintain separate call states itself based on which it should be decided, not some manipulation of phone details.

File: teacher-webapp/src/context/ConferenceContext.js [hunk 3]
```diff
@@ -29,14 +29,36 @@ export const ConferenceProvider = ({ children }) => {
...
+      }
+    });
+  };
+
+  const clearSelectedStudents = () => {
+    setSelectedStudents([]);
   };
 
   const handleSSEEvent = (event) => {
+    console.log("[ConferenceContext] SSE Event received:", {
```
Comment (Soumabha) [inline teacher-webapp/src/context/ConferenceContext.js]: Console logs show up in UI and is best kept only for debugging.

File: teacher-webapp/src/context/ConferenceContext.js [hunk 4]
```diff
@@ -49,6 +71,13 @@ export const ConferenceProvider = ({ children }) => {
         `[ConferenceContext] ${participant.name} (${phoneNumber}): ${previousStatus} -> ${participant.call_status}`
       );
 
+      // Debug: Log raised hand status changes
+      if (participant.is_raised !== undefined) {
```
Comment (Soumabha) [inline teacher-webapp/src/context/ConferenceContext.js]: Is the flag a trigger or state? We need to invert the flag based on multiple key presses.

File: teacher-webapp/src/context/ConferenceContext.js [hunk 5]
```diff
@@ -79,8 +108,20 @@ export const ConferenceProvider = ({ children }) => {
...
     for (let phoneNumber in event.participants) {
+      const participantData = event.participants[phoneNumber];
+      // Map phone_number (snake_case from backend) to phoneNumber (camelCase for frontend)
+      // Use the loop key as the phoneNumber since it's the actual phone number
+      // Ensure boolean values are properly handled (backend might send as string or number)
       const participant = new Participant({
-        ...event.participants[phoneNumber],
+        ...participantData,
+        phoneNumber: phoneNumber, // Use the key from the loop as it's the actual phone number
+        phone_number: participantData.phone_number || phoneNumber, // Also preserve snake_case for compatibility
```
Comment (Soumabha) [inline teacher-webapp/src/context/ConferenceContext.js]: Can we only fix a single type of casing and read from multiple fields if/as required.

File: teacher-webapp/src/pages/ClassroomDetail.js [hunk 6]
```diff
@@ -58,10 +59,28 @@ const ClassroomDetail = () => {
...
 
+  // Clear ALL selected students when classroom changes
+  // Use a ref to track the previous classroomId to avoid sessionStorage dependency
+  useEffect(() => {
+    if (!classroomId) return;
+
+    const hasClassroomChanged =
+      previousClassroomIdRef.current !== null && previousClassroomIdRef.current !== classroomId;
+
+    // Clear ALL selected students when classroom changes (not on initial mount)
```
Comment (Soumabha) [inline teacher-webapp/src/pages/ClassroomDetail.js]: Why not on initial mount?

File: teacher-webapp/src/callPage.js [hunk 7]
```diff
@@ -269,32 +293,99 @@ export function DetailsPage() {
...
+    .map(([normalizedPhone, student]) => {
+      if (!student) return null;
+
+      const userInCall = Array.isArray(userList)
+        ? userList.find(
+            (user) => user && normalizePhoneNumber(getPhoneNumber(user)) === normalizedPhone
+          )
+        : null;
+
+      // If student is in the call (userList), merge all real-time data including raised hand status
```
Comment (Soumabha) [inline teacher-webapp/src/callPage.js]: Can we just maintain a single copy of state for users and use that all across? Whenever the states update, we directly update into that copy itself. That way we dont have to do the merging logic everytime across different places.

File: teacher-webapp/src/components/participants/ParticipantCard.js (hunk 1 above)
Comment (farmanahmed888) [reply teacher-webapp/src/components/participants/ParticipantCard.js]: resolved

File: teacher-webapp/src/context/ConferenceContext.js (hunk 5 above)
Comment (farmanahmed888) [reply teacher-webapp/src/context/ConferenceContext.js]: actually, contentwebapp while registering uses camelCase, and other apps like android app also uses camelCase, we can create a separate ticket where we should migrate all to snake-case and adjust the db, after the demo

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary

The changes focus heavily on centralizing participant state management via a Map (`participantsMap`) in the ConferenceContext, replacing previous reliance on `userList` or merging strategies. Normalization of phone numbers is consistently introduced before API calls and local state updates, improving data consistency. There are improvements to batch participant additions and robust handling in `addParticipants`. UI controls are relabeled for clarity.

### Issues & Recommendations

1. **Redundant/Complex Participant Merging**:
   - The cycle of deriving participants from `studentList`, `userList`, and `conferenceStudents` is replaced by a single source (`participantsMap`), which is good.
   - However, mixing raw/initial data with SSE updates can lead to stale or inconsistent info. Consider strictly separating initial data loading from real-time updates.

2. **Phone Number Normalization**:
   - Ensuring normalization everywhere (+91 prefix + handling nulls) is correct but should be abstracted into a utility function used consistently; currently spread out.

3. **Type Safety & Robustness**:
   - `participant.is_raised` and `participant.is_muted` are mixed boolean/number/string. Use explicit type coercion or validation.
   - `participant.raised_at` handling is inconsistent; ensure it's always a number.

4. **Code Complexity & Duplication**:
   - Multiple `setParticipantsMap` updates could be abstracted to avoid duplicated logic.
   - The SSE event handling updates participants and dispatches notifications, but the logic for detecting disconnections could be clearer.

5. **UI/UX Labeling**:
   - Button labels realigned, correct and clearer. Good.

### Design & Complexity
- Centralized `participantsMap` is a strong pattern, but current update logic is complex; consider more immutable updates or a dedicated reducer.
- The mapping between initial student data, real-time state, and UI display is somewhat convoluted; a normalized, unified participant object model will reduce bugs.

### Score: **75/100**  
Good refactoring effort, but needs cleaner separation of concerns, stricter type handling, and abstraction to reduce complexity and potential bugs.

### Final note
Focus on robust type safety, create utility functions for normalization everywhere, and streamline participant updates with reducers or dedicated update functions.

**Review by a4i-architect — COMMENTED** [review]
### Summary

The main focus of the changes involves centralizing participant state management via `participantsMap` in the `ConferenceContext`, replacing the previous `userList` logic. This improves consistency and real-time updates. Phone number normalization is well integrated but somewhat scattered, especially in API and component layers. Some code, notably in SSE handling and participant filtering, is overly complicated or duplicated.

---

### Issues & Recommendations

1. **Unnecessary Complexity & Duplicates**
   - **Participants handling:** Using both `userList` and `participantsMap` introduces confusion. Synchronize all user data strictly via `participantsMap`.
   - **Filtering students:** The logic for excluding already-in-call students becomes complex due to duplicated normalization. Standardize all phone numbers at data entry to prevent duplicated filtering logic.

2. **Type & Robustness**
   - Ensure all `participant` properties (is_muted, is_raised, call_status) are properly typed/serialized. Handle potential undefined or malformed data more defensively.
   - In `ParticipantCard`, `participant?.is_raised` null checks are good, but consider default values for consistency.

3. **Normalization Approach**
   - Phone number normalization occurs in multiple layers (API, context, utilities). Enforce normalization at input points (e.g., form entries, student toggle) and maintain normalized data throughout.

4. **Code Quality & Readability**
   - Extract participant merging logic in SSE handler into a dedicated function to avoid clutter.
   - Avoid filtering `userList` or `students` multiple times; rely solely on `participantsMap` where possible.

5. **Testing & API**
   - Tests correctly check normalization but lack coverage for edge cases (invalid numbers, undefined fields). Enhance test coverage.
   - API calls normalize numbers consistently, which is good. Confirm all call sites are accounted for.

---

### Design & Cyclomatic Complexity
- Using a single source of truth (`participantsMap`) is a sound pattern, reducing inconsistency.
- SSE handling is overly verbose; refactoring into dedicated functions would improve maintainability.
- The complex filtering and merging logic could be simplified by ensuring normalized data upfront, reducing run-time filtering.

### Score: 75/100
- Solid core shift towards state centralization and normalization.
- Needs cleanup of duplicated logic, clearer separation of concerns, and consistency in data flow.

**Next steps:** Enforce normalization at data ingress, refactor SSE update logic, and simplify participant filters for cleaner, more maintainable code.

**Review by a4i-architect — COMMENTED** [review]
**Summary of Key Feedback**

- **Participant State Management**: Transition from `userList` to a centralized `participantsMap` is good but incomplete. Proper merging of SSE data, roles, and metadata is inconsistent; ensure normalization and merging are robust, especially handling participant properties like `name`, `call_status`, and `is_muted`.

- **Normalization & Consistency**: Several parts normalize phone numbers before API calls and comparisons. This is good. However, ensure normalization is always applied in all code paths (e.g., `handleStudentToggle`, `useConference` hooks). Also, consistently store normalized `phoneNumber`s in state.

- **Redundant/Obsolete Code**: Removed `userList`, but some code still references it (even if commented out). Clean fully to avoid confusion.

- **Filtering & Participant Inclusion**: The approach to filtering offered students is improved with centralized participants info. However, ensure `getAllParticipants()` and `getParticipants()` functions are correctly wired and provide up-to-date data for UI logic.

- **Edge Cases**:
  - `ParticipantCard` now guards `participant` being null.
  - `ClassroomDetail` now normalizes phone numbers when toggling students—good practice.
  - When adding participants, check for invalid or duplicate entries.

- **Concurrency & Performance**:
  - Batch `addParticipant` calls with `Promise.all` for efficiency.
  - Consider debouncing toggles or batch updates to avoid rapid state changes.

- **Testing & Consistency**:
  - Tests correctly verify normalization (adding `"91"` prefix). 
  - Confirm test data and test API responses match expected format.

- **Design & Complexity**: 
  - Transition to a `participantsMap` improves clarity but increases complexity in merging and updating logic.
  - Necessary to streamline participant state sync from SSE events, avoiding race conditions or stale data.

**Overall Score:**  
**78/100** — Good progress, but needs cleaner merging logic, consistent normalization, and more comprehensive state management. Do refactor for clarity and robustness before acceptance.

**Review by a4i-architect — COMMENTED** [review]
### High-Level Summary
The major updates focus on centralizing participant state management with a participantsMap in the context, replacing the previous userList-based approach. Phone number normalization is consistently applied across API calls, components, and state updates to ensure data integrity. Several improvements include better handling of dynamic participant data, filtering, and avoiding duplication.

### Key Issues & Recommendations
1. **Redundant Participant State**: Complex merging of conferenceStudents and userList is replaced by a single participantsMap, but some components (e.g., callPage.js) still mix conceptual models. Consistently use the centralized map everywhere.
2. **Phone Number Normalization**: Properly normalize all phone inputs (UI, API, SSE) — this addresses bugs/issues in participant matching.
3. **Component State & Effects**: Some components still manage their own user/list state (callPage.js, ClassroomeDetail.js). Officially migrate to using the context's `participantsMap` for consistency.
4. **Participant Filtering**: Enhanced filtering logic avoids duplicates, checks if students are already in call based on the centralized map, improving robustness.
5. **API Calls & Testing**: All API interactions now normalize phone numbers. Tests are correctly updated; ensure they cover all code paths (e.g., invalid phones). Confirm normalization function is robust.
6. **Code Cleanliness**: Remove commented/unnecessary code (original useEffect on `userList`), refactor duplicated logic. Maintain clear separation of concerns.
7. **Design**: Using a Map in context is good; ensure frontend components avoid inconsistent partial states.

### Design & Complexity
- Overall, the approach is sound: central state management via a Map reduces duplication and stale data.
- Complexity is acceptable; offset with consistent normalization.
- Watch out for asynchronous race conditions if multiple components update participants state simultaneously.

### Score: 75/100
- Strong move toward centralized participant management and phone normalization.
- Minor issues: component logic can be further refactored for clarity, and some residual state handling could be cleaned up. Overall, the pattern is sound for production.

**Next Steps**:
- Fully migrate all participant references to context state.
- Add tests validating normalization and participant filtering.
- Review component re-renders caused by context updates.

**References**:
- State normalization best practices: https://redux.js.org/recipes/structuring-reducers#normalize-shape-of-state
- Phone number normalization strategies: https://libphonenumber.github.io/libphonenumber/
- React state management and context: https://reactjs.org/docs/context.html

File: teacher-webapp/src/context/ConferenceContext.js (hunk 5 above)
Comment (farmanahmed888) [reply teacher-webapp/src/context/ConferenceContext.js]: https://github.com/A4i-tech/.github/issues/189

**Review by Soumabha — APPROVED** [review]
